docs/TEMPORARY_DIRECTORIES.md

```diff
@@ -0,0 +1,218 @@
+---
+title: Using /tmp/ And /var/tmp/ Safely
+---
+
+# Using `/tmp/` And `/var/tmp/` Safely
+
+`/tmp/` and `/var/tmp/` are two world-writable directories Linux systems
+provide for temporary files. The former is typically on `tmpfs` and thus
+backed by RAM/swap, and flushed out on each reboot. The latter is typically a
+proper, persistent file system, and thus backed by physical storage. This
+means:
+
+1. `/tmp/` should be used for smaller, size-bounded files only; `/var/tmp/` should be used for everything else.
+
+2. Data that shall survive a boot cycle shouldn't be placed in `/tmp/`.
+
+If the `$TMPDIR` environment variable is set, use that path, and neither use
+`/tmp/` nor `/var/tmp/` directly.
+
+See
+[file-hierarchy(7)](https://www.freedesktop.org/software/systemd/man/file-hierarchy.html)
+for details about these two (and most other) directories of a Linux system.
+
+## Common Namespace
+
+Note that `/tmp/` and `/var/tmp/` each define a common namespace shared by all
+local software. This means guessable file or directory names below either
+directory directly translate into a 🚨 Denial-of-Service (DoS) 🚨 vulnerability
+or worse: if some software creates a file or directory `/tmp/foo` then any
+other software that wants to create the same file or directory `/tmp/foo`
+either will fail (as the file already exists) or might be tricked into using
+untrusted files. Hence: do note use guessable names in `/tmp/` or `/var/tmp/` —
+if you do you open yourself up to a local DoS exploit or worse. (You can get
+away with using guessable names, if you pre-create subdirectories below `/tmp/`
+for them, like X11 does with `/tmp/.X11-unix/` through `tmpfiles.d/`
+drop-ins. However this is not recommended, as it is fully safe only if these
+directories are pre-created during early boot, and thus problematic if package
+installation during runtime is permitted.)
+
+To protect yourself against these kinds of attacks Linux provides a couple of
+APIs that help you avoiding guessable names. Specifically:
+
+1. Use [`mkstemp()`](http://man7.org/linux/man-pages/man3/mkstemp.3.html)
+(POSIX), `mkostemp()` (glibc),
+[`mkdtemp()`](http://man7.org/linux/man-pages/man3/mkdtemp.3.html) (POSIX),
+[`tmpfile()`](http://man7.org/linux/man-pages/man3/tmpfile.3.html) (C89)
+
+2. Use [`open()`](http://man7.org/linux/man-pages/man2/open.2.html) with `O_TMPFILE` (Linux)
+
+3. [`memfd_create()`](http://man7.org/linux/man-pages/man2/memfd_create.2.html)
+   (Linux; this doesn't bother with `/tmp/` or `/var/tmp/` at all, but uses
+   the same RAM/swap backing as `tmpfs` uses, hence is very similar to `/tmp/`
+   semantics.
+
+For system services systemd provides the `PrivateTmp=` boolean setting. If
+turned on for a service (👍 which is highly recommended), `/tmp/` and
+`/var/tmp/` are replaced by private non-guessable sub-directories, implemented
```

Comment:
I'd drop "non-guessable" here. I know what you mean, but how exactly systemd implements this is a technical detail that people don't care about. And also, once the directory has been created, anyone can see the name using `ls -l /tmp/systemd-private* -d`.